merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,872 @@
+        }
+
+        Ok(None)
+    }
+
+    // Returns a first available element.
+    pub fn first(&self, sequence_number: usize) -> Result<Option<&mut HashSetCell>, HashSetError> {
+        for i in
+            0..usize::try_from(self.capacity_values).map_err(|_| HashSetError::IntegerOverflow)?
+        {
+            let value_bucket = unsafe { &mut *self.values.as_ptr().add(i) };
+
+            if let Some(value_bucket) = value_bucket {
+                if let Some(other_sequence_number) = value_bucket.sequence_number {
+                    let sequence_diff = sequence_number.checked_sub(other_sequence_number);
+                    if let Some(sequence_diff) = sequence_diff {
+                        if sequence_diff < self.sequence_threshold {
+                            return Ok(Some(value_bucket));
+                        }
+                    }
+                } else {
+                    return Ok(Some(value_bucket));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    // Check if the hash set contains a value.
+    pub fn contains(&self, value: &BigUint, sequence_number: usize) -> Result<bool, HashSetError> {
+        for i in
+            0..usize::try_from(self.capacity_values).map_err(|_| HashSetError::IntegerOverflow)?
+        {
+            let i = I::try_from(i).map_err(|_| HashSetError::IntegerOverflow)?;
+            let probe_index = (value.clone() + i.to_biguint().unwrap() * i.to_biguint().unwrap())
+                % self.capacity_values.to_biguint().unwrap();
+            let probe_index = probe_index.to_usize().unwrap();
+            let index_bucket = unsafe { &*self.indices.as_ptr().add(probe_index) };
+
+            match index_bucket {
+                Some(value_index) => {
+                    let value_bucket = unsafe {
+                        &mut *self.values.as_ptr().add(
+                            usize::try_from(*value_index).map_err(|_| HashSetError::UsizeConv)?,
+                        )
+                    };
+
+                    if let Some(value_bucket) = value_bucket {
+                        if let Some(other_sequence_number) = value_bucket.sequence_number {
+                            let sequence_diff = sequence_number.checked_sub(other_sequence_number);
+                            if let Some(sequence_diff) = sequence_diff {
+                                if sequence_diff >= self.sequence_threshold {
+                                    return Ok(false);
+                                }
+                            }
+                        }
+
+                        if &value_bucket.value == value {
+                            return Ok(true);
+                        }
+                    }
+                }
+                None => {
+                    return Ok(false);
+                }
+            }
+        }
+
+        Ok(false)
+    }
+
+    // Marks the given element with a given sequence number.
+    pub fn mark_with_sequence_number(
```

Comment:
When will this function be called?
Can't we just pass the index of the element?

Reply:
In adress Merkle tree update, where we get the `first()` element and then use `mark_with_sequence_number()`.

Yeah, passing the index of the element is an another option and probably better for nullifying UTXOs.